### ISSUE TITLE
[Feature]: Implement radial quantization (TurboQuant)

### ISSUE THREAD

**Issue by Mec-iS** [Mec-iS/arrowspace-rs#106]
## Polar Coordinates \& the Rayleigh Pipeline

Polar coordinates are **fully compatible** with Rayleigh scoring. The squared L₂ difference between a pair of polar-encoded dimensions reduces via the law of cosines to:

$$ (r_a\cos\theta_a - r_b\cos\theta_b)^2 + (r_a\sin\theta_a - r_b\sin\theta_b)^2 = r_a^2 + r_b^2 - 2r_ar_b\cos(\Delta\theta_{ab}) $$

At 4-bit angle resolution (16 sectors), $\cos(\Delta\theta)$ becomes a **16-entry integer lookup table** . The Rayleigh numerator $x^T L x = \sum_{(i,j)\in E} w_{ij}(x_i - x_j)^2$ therefore becomes a sum of table lookups scaled by float radii — purely integer arithmetic on the centroid side, compatible with Rayon's parallel iterators that already power `nodeEnergyAndDispersion` [^3][^1].

For the **Gini dispersion term**, angular spread $\Delta\theta = |\theta_i - \theta_j|$ is a natural integer-arithmetic proxy for directional dispersion between neighbouring centroids, replacing the float64 L₂-difference computation in the current implementation [^1].

TurboQuant's 1-bit QJL residual is critical for λ correctness: PolarQuant alone introduces a small bias from 22.5° angle-grid discretisation. The residual correction restores **unbiased Rayleigh estimation**, meaning λ values at 5-bit precision remain statistically indistinguishable from float64 baselines — directly protecting the CVE benchmark MRR=0.75 result at eta=0.5, steps=4 [^2].

### Proposed Solution

implement polar quatization for Rayleigh

### Priority

Medium - Would be helpful

### Estimated Complexity

Unknown - Not sure about implementation

### Feature Type

- [ ] API enhancement
- [x] Performance improvement
- [ ] New functionality
- [ ] Developer experience improvement
- [ ] Documentation enhancement
- [ ] Configuration option
- [ ] Integration with external service

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct
- [x] I have searched existing issues to avoid duplicates
- [ ] I am willing to help implement this feature if guidance is provided

**Comment by Mec-iS** [comment]
### 1. ArrowSpace-Specific improvements
The document highlights three major algorithmic integrations tailored directly to ArrowSpace's architecture:

*   **QJL for Item Index Compression & Cosine Search:** ArrowSpace's hybrid search equation (`score_i = α · cosine(q, x_i) + (1-α) · proximity(λ_q, λ_i)`) relies heavily on cosine similarity. QJL’s asymmetric estimator allows ArrowSpace to store indexed items as 1-bit sketches plus their norm, while evaluating full-precision queries online. This yields a 5x+ memory reduction with provably unbiased cosine estimates. 
*   **PolarQuant for Persistent Storage:** ArrowSpace already uses Johnson-Lindenstrauss (JL) random projection (`reduction.rs`), which naturally preconditions the vectors into the near-Gaussian state required by PolarQuant. When serializing the index to Parquet (`storage/`), vectors can be converted to polar coordinates where the angles are quantized to 8-bit and the radius is kept at full precision, achieving a 4x size reduction on disk. 
*   **TurboQuant for Rayleigh Computation (λ):** The computation of the Rayleigh quotient ($\lambda_i = \frac{x_i^T L x_i}{x_i^T x_i}$) is a bilinear form structurally identical to transformer attention logits. TurboQuant applies a two-stage approach: an MSE-optimal scalar quantizer (PolarQuant) for the main signal, and a 1-bit QJL transform on the residual to correct bias. This allows ArrowSpace to compute unbiased $\lambda$ estimates using 3–4 bit item representations, achieving up to an 8x throughput speedup for large $N$ batches. 

***

### 2. Analysis of Proposed Improvements
The integrations align perfectly with ArrowSpace's goal of scaling to extreme high-dimensional latent spaces while preserving the mathematical purity of the graph wiring.
*   **Zero-Overhead Compatibility:** All three algorithms avoid per-block normalization constants (scale/zero-point). They rely instead on the shared JL projection matrix already present in ArrowSpace, fitting seamlessly into the "single numeric contract" invariant. 
*   **The Unbiased Guarantee:** Quantizing variables naively would normally corrupt the topological properties of the Graph Laplacian. TurboQuant’s bias correction guarantees that the structural smoothness/roughness encoded in $\lambda$ remains a mathematically valid proxy for the Dirichlet energy, even at 3-4 bits. 

***

### 3. Drafted Implementation Plan
Given your instruction, **Phase 1 prioritizes the TurboQuant implementation on the Rayleigh computation**.

#### Phase 1: Quantized Rayleigh Computation (TurboQuant) *[Highest Priority]*
**Objective:** Compute $\lambda_i$ securely and efficiently over 3-4 bit compressed vectors.
1.  **TurboQuant Structs:** Create a new `TurboQuantEstimator` module in `arrowspace`.
2.  **Stage 1 (Main Signal):** Implement the MSE-optimal scalar quantizer via recursive polar transformation on the projected vectors. Group coordinates into pairs, extract radii, and quantize the independent angles.
3.  **Stage 2 (Residual Bias Correction):** Extract the quantization error (residual) and apply a 1-bit QJL transform to it.
4.  **Rayleigh Integration:** Refactor the $\lambda$ computation step (e.g., inside `prepare_query_item` and graph construction). Instead of full-precision dot products for $x^T L x$, use the TurboQuant unbiased inner product estimator combining the main signal and the 1-bit residual.

#### Phase 2: QJL for In-Memory Vector Search
**Objective:** Shrink RAM footprint for `search_lambda_aware` cosine scoring.
1.  Introduce a `QJLArrowIndex` trait/struct.
2.  Modify the `ImplicitProjection` pipeline so that after multiplying by the JL matrix $R$, we apply `sign()` to store vectors as bit-arrays (plus an `f32` norm scalar).
3.  Rewrite the `cosine_sim` tie-breaker in `search_energy` and `search_lambda_aware` to utilize the QJL asymmetric distance estimator (Hamming distance mapping for the 1-bit keys $\times$ full precision query).

#### Phase 3: PolarQuant for Persistent Parquet Storage
**Objective:** Reduce disk size of serialized ArrowSpace indices.
1.  Intercept vectors in `storage/parquet.rs` before writing `DenseMatrix`.
2.  Apply recursive polar transforms to the feature coordinates (taking advantage of the existing JL projection).
3.  Serialize the 8-bit quantized angles and the single full-precision radius.
4.  Implement the inverse polar transform on `ArrowSpace::new_from_storage`.

**Comment by Mec-iS** [comment]
Here is the concrete implementation plan based on your constraints. We will stick to pure Rust (avoiding FFI/CUDA), handle the power-of-two requirement dynamically during the JL projection, and keep the `DenseMatrix<f64>` fully intact while using the compressed representation strictly as an acceleration cache for $\lambda$ computations and search.

### Phase 1: Dynamic Power-of-Two Projection (`src/reduction.rs`)
To satisfy PolarQuant's recursive polar transform requirements without mutating the user's raw dataset, we will enforce the power-of-two constraint strictly during the Johnson-Lindenstrauss (JL) projection step.

*   **Implementation Steps:**
    1.  Update the `ImplicitProjection::new()` method to calculate the next power of two for the target dimension $r$: `let r_padded = r.next_power_of_two();`
    2.  Generate the JL random matrix $R$ with dimensions $r_{padded} \times F$ instead of $r \times F$. 
    3.  Because the JL transform is just random Gaussian/Rademacher projections, extending $r$ slightly strictly *improves* the Johnson-Lindenstrauss bounds. It requires zero zero-padding of the input data; it just creates a slightly wider projected vector that satisfies the $2^n$ constraint perfectly.

### Phase 2: Native Rust TurboQuant Module (`src/quant/turboquant.rs`)
Instead of injecting complex FFI, we will build a native Rust module referencing the available crates (`turbo-quant`/`turboquant`) or implementing the two-stage algorithm natively using `std::simd` and `rayon`.

*   **Implementation Steps:**
    1.  **Stage 1 (PolarQuant):** Implement `RecursivePolarTransform`. Given a projected vector of size $2^n$, group pairs, compute local radii and angles, and quantize the independent angles to 8-bit. Store the single global radius as `f32`.
    2.  **Stage 2 (QJL Residual):** Compute the reconstruction error (residual) of Stage 1. Apply a 1-bit `sign()` transform to this residual.
    3.  **Asymmetric Estimator:** Implement the unbiased dot product function: `dot_asymmetric(query_full_precision, item_quantized)`. This combines the full-precision query with the 8-bit PolarQuant signal and the 1-bit QJL residual correction.

### Phase 3: The Quantized Rayleigh Cache (`src/core.rs` & `src/builder.rs`)
The full raw dataset is preserved for tasks that strictly require it, while the quantized versions act as a side-car structure to accelerate $O(N)$ operations.

*   **Implementation Steps:**
    1.  Create a new struct:
        ```rust
        pub struct QuantizedCache {
            pub polar_angles: Vec<Vec<u8>>, // 8-bit quantized angles
            pub residuals: Vec<Vec<bool>>,  // 1-bit QJL signs
            pub radii: Vec<f32>,            // Global norm scalar
        }
        ```
    2.  Modify `ArrowSpace`: Add `pub quantized_cache: Option<QuantizedCache>` alongside `pub data: DenseMatrix<f64>`.
    3.  Modify `ArrowSpaceBuilder::build()`: After the data matrix is stored and the projection matrix is finalized, `.into_par_iter()` over the items to generate the `QuantizedCache` in parallel.

### Phase 4: Refactoring Rayleigh ($\lambda$) Computation (`src/graph.rs` & `src/search/`)
The Rayleigh quotient requires evaluating $x^T L x$. Because $L$ is a sparse graph Laplacian built over features, evaluating this is structurally a series of dot products.

*   **Implementation Steps:**
    1.  **Batch $\lambda$ Calculation:** During graph construction or index prep, replace the full-precision dense vector multiplications with the `dot_asymmetric` estimator.
    2.  **Query Preparation (`prepare_query_item`):** When a new search query $q$ arrives, it remains in `f64`/`f32`. Compute $\lambda_q$ by mapping $q$ against the Laplacian $L$, utilizing the quantized cache if the computation involves the stored item vectors.
    3.  **Hybrid Search Tie-Breakers:** In `search_lambda_aware` and `search_energy`, update the cosine similarity tie-breaker to use the ultra-fast `dot_asymmetric` method between the full-precision query $q$ and the `QuantizedCache`, drastically reducing memory bandwidth during top-k evaluation.

### Phase 5: Testing and Integration (`tests/test_quantization.rs`)
Because this touches the mathematical core, we need to guarantee that the TurboQuant unbiased estimator doesn't degrade the existing search accuracy.

*   **Implementation Steps:**
    1.  Create tests comparing $\lambda_{exact}$ (using `DenseMatrix<f64>`) vs $\lambda_{quantized}$ (using `QuantizedCache`).
    2.  Assert that the variance between the two is bound by the expected TurboQuant distortion limits.
    3.  Benchmark memory consumption and throughput using `criterion`, expecting a ~4x-8x speedup on batch Rayleigh execution without GPU overhead.

**Comment by Mec-iS** [comment]
implement quantization as an optional sidecar phase after JL projection and before the IndexScorePhase, while keeping the existing `DenseMatrix<f64>` path exact and using the quantized path only for fast Rayleigh/TauMode and later lambda-aware search.  This fits the current ArrowSpace design, which already separates raw data, projection metadata, lambdas, centroid metadata, and search-oriented state instead of collapsing everything into one representation.[^1]

## Placement

The repo already exposes a staged pipeline through `EigenMaps`: clustering, Laplacian construction, TauMode lambda computation, optional spectral feature-graph construction, and search.  So the clean design is to add a `QuantizationPhase` between projection/compression and `IndexScorePhase`, which also matches the modular “Graph Wiring” direction where the last phase stays pluggable by capability.[^1]

For the first pass, scope quantization to `TauModeScore` only. Current query preparation already projects queries when needed and computes synthetic lambda, while lambda-aware search blends cosine similarity with lambda proximity, so Rayleigh/TauMode is the highest-value insertion point with the smallest surface area.[^1]

## API shape

```
ArrowSpace today keeps raw data as `DenseMatrix<f64>` and stores projection info, lambdas, sorted lambdas, centroid maps, subcentroids, and item norms alongside it.  Add `quantized_cache: Option<QuantizedCache>` next to those fields, not instead of them, so exact paths remain available for analysis, search debugging, and regression tests.[^1]
```

```rust
pub enum QuantizationMode {
    None,
    PolarRayleigh {
        angle_bits: u8,      // e.g. 4 => 16 sectors
        use_qjl_residual: bool,
        qjl_dim: usize,
        pad_to_pow2: bool,
    },
}

pub trait QuantizationPhase: Send + Sync {
    type Code;
    fn encode(&self, x: &[f64]) -> Self::Code;
    fn estimate_dot(&self, q: &[f64], code: &Self::Code) -> f64;
    fn estimate_sq_l2(&self, a: &Self::Code, b: &Self::Code) -> f64;
}

pub struct QuantizedCache {
    pub codes: Vec<PolarCode>,
    pub padded_dim: usize,
    pub angle_bits: u8,
}
```

I would keep the MVP simple: flat 2D-pair polar encoding for the projected vector, not the full recursive tree in the first PR. That already gives you the law-of-cosines lookup for $(x_i-x_j)^2$ and keeps the implementation small enough to validate before adding deeper TurboQuant machinery.

## Quantizer choice

PolarQuant is a strong fit because it uses random preconditioning/projection, converts vectors to polar coordinates, and quantizes angles without per-block normalization overhead.  TurboQuant’s second stage then applies a 1-bit QJL transform to the residual, giving an unbiased inner-product estimator instead of the biased score from MSE-only quantization.[^2][^3][^4][^5][^6][^7]

For ArrowSpace specifically, I would map that into two layers:

- **Base signal:** quantized polar code over projected item vectors.
- **Bias correction:** optional QJL residual sketch used only in Rayleigh/dot estimation paths where λ ranking quality matters most.

There are already Rust APIs documenting `encode` and `inner_product_estimate` style TurboQuant/PolarQuant interfaces, so you can prototype the estimator contract quickly even if you later replace the backend with a native in-repo implementation.[^8][^9][^10]

## Integration points

- `reduction.rs`: when quantization is enabled, pad the reduced dimension to `next_power_of_two` before producing the projected representation. The current code already treats projection metadata as first-class state and query preparation depends on it.[^1]
- `builder.rs` / `EigenMaps`: after projection and before `computetaumode`, build `QuantizedCache` in parallel and store it on `ArrowSpace`. The current pipeline already has explicit build stages and parallel TauMode computation, so this is the natural insertion point.[^1]
- `search/taumode.rs`: add a quantized estimator path for the Rayleigh numerator and dispersion term behind a config flag, while preserving the exact path as fallback. Current TauMode code computes synthetic lambdas in parallel from bounded Rayleigh energy plus dispersion.[^1]
- `search`: leave cosine tie-break quantization for a second PR. Current lambda-aware search already uses cosine as the semantic term layered with lambda proximity, so it can be upgraded later without blocking the Rayleigh work.[^1]

Concretely, I would add:

1. `with_quantization(QuantizationMode)` on `ArrowSpaceBuilder`.
2. `build_quantized_cache(&mut self)` as a reusable phase.
3. `compute_synthetic_lambda_quantized(...)` alongside the exact TauMode path.
4. A capability builder such as `build_search(TauModeScore, QuantizationMode)` so quantization remains modular rather than hardwired.

## Validation

Keep the first PR narrow, which matches the project guidance to make minimum, surgical changes and verify them with focused tests.  Add tests that compare exact vs quantized lambda on the same indexed items, preserve existing scale-invariance and projection-consistency properties, and enforce a ranking-drift budget instead of only mean absolute error; the current test suite already checks deterministic query lambda computation, scale invariance, and projection consistency in TauMode-related paths.[^11][^1]

My recommended first PR is:

- Add `QuantizationMode`, `QuantizedCache`, and builder wiring.
- Implement flat pairwise polar quantization with a cosine lookup table.
- Add a quantized Rayleigh/TauMode path only.
- Benchmark exact vs quantized λ throughput and top-k ranking drift.
